File: frontend/src/types/workflow.ts
// 工作流相关类型，与后端 Go 结构对齐
// 字段命名遵循后端 JSON tag（snake_case）

export interface Position {
  x: number;
  y: number;
}

// 节点实例（工作流中的具体节点）
// 注意：不带 stage 字段，阶段归属由可达性导出
// 使用 type 而非 interface，满足 React Flow Node<T extends Record<string, unknown>> 约束
export type NodeInstance = {
  instance_id: string; // UUID
  type_id: string;
  config: Record<string, unknown>;
  state?: string;
  error_msg?: string;
  position: Position;
};

// 连线端口引用
export interface PortRef {
  node: string; // 引用 NodeInstance.instance_id
  port: string; // PortDef.id
}

export interface EdgeConfig {
  from: PortRef;
  to: PortRef;
}

// VariableDef 工作流/集合内部变量定义
export interface VariableDef {
  name: string;
  var_type: string; // 复用 PortType 字符串
  default?: unknown;
}

// 工作流定义（顶层结构，与后端 core.WorkflowDef 对齐）
export interface WorkflowDef {
  id: string;
  name: string;
  description: string;
  variables: VariableDef[];
  nodes: NodeInstance[];
  edges: EdgeConfig[];
}

File: internal/core/workflow.go
package core

// UPDATE 阶段配置
type UpdateConfig struct {
	DeltaType    string `json:"delta_type"    toml:"delta_type"` // interval|cron|manual
	DeltaSeconds int64  `json:"delta_seconds" toml:"delta_seconds"`
	CronExpr     string `json:"cron_expr,omitempty" toml:"cron_expr"`
}

// 连线定义（持久化格式）
type EdgeConfig struct {
	From PortRef `json:"from" toml:"from"`
	To   PortRef `json:"to"   toml:"to"`
}

// 端口引用（节点实例ID + 端口ID）
type PortRef struct {
	Node string `json:"node" toml:"node"`
	Port string `json:"port" toml:"port"`
}

// VariableDef 工作流/集合内部变量定义
type VariableDef struct {
	Name    string   `json:"name"     toml:"name"`
	VarType PortType `json:"var_type" toml:"var_type"`
	Default any      `json:"default,omitempty" toml:"default,omitempty"`
}

// WorkflowDef 工作流定义（持久化到 TOML）
// 注意：
//   - 无 UpdateConfig 字段（Delta 配置移到 system_update 节点的 Config 中）
//   - 无 SubWorkflows 字段（子工作流通过 sub_workflow_call 节点表达）
type WorkflowDef struct {
	ID          string         `json:"id"          toml:"id"`
	Name        string         `json:"name"        toml:"name"`
	Description string         `json:"description" toml:"description"`
	Variables   []VariableDef  `json:"variables"   toml:"variables"`
	Nodes       []NodeInstance `json:"nodes"       toml:"nodes"`
	Edges       []EdgeConfig   `json:"edges"       toml:"edges"`
}
